README.md
# markdown-it-abbr

[![CI](https://github.com/markdown-it/markdown-it-abbr/actions/workflows/ci.yml/badge.svg)](https://github.com/markdown-it/markdown-it-abbr/actions/workflows/ci.yml)
[![NPM version](https://img.shields.io/npm/v/markdown-it-abbr.svg?style=flat)](https://www.npmjs.org/package/markdown-it-abbr)
[![Coverage Status](https://img.shields.io/coveralls/markdown-it/markdown-it-abbr/master.svg?style=flat)](https://coveralls.io/r/markdown-it/markdown-it-abbr?branch=master)

> Abbreviation (`<abbr>`) tag plugin for [markdown-it](https://github.com/markdown-it/markdown-it) markdown parser.

__v1.+ requires `markdown-it` v4.+, see changelog.__

Markup is based on [php markdown extra](https://michelf.ca/projects/php-markdown/extra/#abbr) definition, but without multiline support.

Markdown:

```
*[HTML]: Hyper Text Markup Language
*[W3C]:  World Wide Web Consortium
The HTML specification
is maintained by the W3C.
```

HTML:

```html
<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification
is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>.</p>
```

## Install

node.js, browser:

```bash
npm install markdown-it-abbr --save
bower install markdown-it-abbr --save
```

## Use

```js
var md = require('markdown-it')()
            .use(require('markdown-it-abbr'));

md.render(/*...*/) // see example above
```

_Differences in browser._ If you load script directly into the page, without
package system, module will add itself globally as `window.markdownitAbbr`.


## License

[MIT](https://github.com/markdown-it/markdown-it-abbr/blob/master/LICENSE)


CHANGELOG.md
2.0.0 / 2023-12-03
------------------

- Rewrite to ESM.
- Remove `dist/` from repo (build on package publish).


1.0.4 / 2016-08-31
------------------

- Fixed infinite loop caused by empty abbreviations like `*[]: foo`. Empty abbreviations are no longer parsed.


1.0.3 / 2016-01-05
------------------

- Fixed uc.micro use.


1.0.2 / 2015-12-31
------------------

- maintenance, bower files rebuild


1.0.1 / 2015-12-31
------------------

- Resolved collision with `linkify-it`.
- Extended list of start/stop punctuation characters to unicode ones.


1.0.0 / 2015-03-12
------------------

- Markdown-it 4.0.0 support. Use previous version for 2.x-3.x.


0.1.0 / 2015-01-04
------------------

- First release.


test/test.js
'use strict';


var path     = require('path');
var generate = require('markdown-it-testgen');

/*eslint-env mocha*/

describe('markdown-it-abbr', function () {
  var md = require('markdown-it')({ linkify: true })
              .use(require('../'));

  generate(path.join(__dirname, 'fixtures/abbr.txt'), md);
});


test/test.mjs
import { fileURLToPath } from 'node:url'
import markdownit from 'markdown-it'
import generate from 'markdown-it-testgen'

import abbr from '../index.mjs'

describe('markdown-it-abbr', function () {
  const md = markdownit({ linkify: true }).use(abbr)

  generate(fileURLToPath(new URL('fixtures/abbr.txt', import.meta.url)), md)
})


index.mjs
// Enclose abbreviations in <abbr> tags
//
export default function abbr_plugin (md) {
  const escapeRE        = md.utils.escapeRE
  const arrayReplaceAt  = md.utils.arrayReplaceAt

  // ASCII characters in Cc, Sc, Sm, Sk categories we should terminate on;
  // you can check character classes here:
  // http://www.unicode.org/Public/UNIDATA/UnicodeData.txt
  const OTHER_CHARS      = ' \r\n$+<=>^`|~'

  const UNICODE_PUNCT_RE = md.utils.lib.ucmicro.P.source
  const UNICODE_SPACE_RE = md.utils.lib.ucmicro.Z.source

  function abbr_def (state, startLine, endLine, silent) {
    let labelEnd
    let pos = state.bMarks[startLine] + state.tShift[startLine]
    const max = state.eMarks[startLine]

    if (pos + 2 >= max) { return false }

    if (state.src.charCodeAt(pos++) !== 0x2A/* * */) { return false }
    if (state.src.charCodeAt(pos++) !== 0x5B/* [ */) { return false }

    const labelStart = pos

    for (; pos < max; pos++) {
      const ch = state.src.charCodeAt(pos)
      if (ch === 0x5B /* [ */) {
        return false
      } else if (ch === 0x5D /* ] */) {
        labelEnd = pos
        break
      } else if (ch === 0x5C /* \ */) {
        pos++
      }
    }

    if (labelEnd < 0 || state.src.charCodeAt(labelEnd + 1) !== 0x3A/* : */) {
      return false
    }

    if (silent) { return true }

    const label = state.src.slice(labelStart, labelEnd).replace(/\\(.)/g, '$1')
    const title = state.src.slice(labelEnd + 2, max).trim()
    if (label.length === 0) { return false }
    if (title.length === 0) { return false }
    if (!state.env.abbreviations) { state.env.abbreviations = {} }
    // prepend ':' to avoid conflict with Object.prototype members
    if (typeof state.env.abbreviations[':' + label] === 'undefined') {
      state.env.abbreviations[':' + label] = title
    }

    state.line = startLine + 1
    return true
  }

  function abbr_replace (state) {
    const blockTokens = state.tokens

    if (!state.env.abbreviations) { return }

    const regSimple = new RegExp('(?:' +
      Object.keys(state.env.abbreviations).map(function (x) {
        return x.substr(1)
      }).sort(function (a, b) {
        return b.length - a.length
      }).map(escapeRE).join('|') +
    ')')

    const regText = '(^|' + UNICODE_PUNCT_RE + '|' + UNICODE_SPACE_RE +
                    '|[' + OTHER_CHARS.split('').map(escapeRE).join('') + '])' +
            '(' + Object.keys(state.env.abbreviations).map(function (x) {
      return x.substr(1)
    }).sort(function (a, b) {
      return b.length - a.length
    }).map(escapeRE).join('|') + ')' +
            '($|' + UNICODE_PUNCT_RE + '|' + UNICODE_SPACE_RE +
                    '|[' + OTHER_CHARS.split('').map(escapeRE).join('') + '])'

    const reg = new RegExp(regText, 'g')

    for (let j = 0, l = blockTokens.length; j < l; j++) {
      if (blockTokens[j].type !== 'inline') { continue }
      let tokens = blockTokens[j].children

      // We scan from the end, to keep position when new tags added.
      for (let i = tokens.length - 1; i >= 0; i--) {
        const currentToken = tokens[i]
        if (currentToken.type !== 'text') { continue }

        let pos = 0
        const text = currentToken.content
        reg.lastIndex = 0
        const nodes = []

        // fast regexp run to determine whether there are any abbreviated words
        // in the current token
        if (!regSimple.test(text)) { continue }

        let m

        while ((m = reg.exec(text))) {
          if (m.index > 0 || m[1].length > 0) {
            const token = new state.Token('text', '', 0)
            token.content = text.slice(pos, m.index + m[1].length)
            nodes.push(token)
          }

          const token_o = new state.Token('abbr_open', 'abbr', 1)
          token_o.attrs = [['title', state.env.abbreviations[':' + m[2]]]]
          nodes.push(token_o)

          const token_t = new state.Token('text', '', 0)
          token_t.content = m[2]
          nodes.push(token_t)

          const token_c = new state.Token('abbr_close', 'abbr', -1)
          nodes.push(token_c)

          reg.lastIndex -= m[3].length
          pos = reg.lastIndex
        }

        if (!nodes.length) { continue }

        if (pos < text.length) {
          const token = new state.Token('text', '', 0)
          token.content = text.slice(pos)
          nodes.push(token)
        }

        // replace current node
        blockTokens[j].children = tokens = arrayReplaceAt(tokens, i, nodes)
      }
    }
  }

  md.block.ruler.before('reference', 'abbr_def', abbr_def, { alt: ['paragraph', 'reference'] })

  md.core.ruler.after('linkify', 'abbr_replace', abbr_replace)
};


index.js
// Enclose abbreviations in <abbr> tags
//
'use strict';

/**
* @param {*} md The markdown-it plugin instance.
* @param {Object=} abbrDefList A list of abbreviations and their definitions 
*   {"HTML": "Hyper Text Markup Language", "W3C": "World Wide Web Consortium"}. 
*   It will be merged with the reference style abbreviation definitions like `*[HTML]: Hyper Text Markup Language` 
*   inside the markdown files (file definitions overwrite existing list definitions by default).
* @param {boolean=} [listPriorsFile=false] If false (default) definitions inside the markdown file overwrite 
*   existing list definitions. If true list definitions overwrite existing definitions in the markdown file. 
*/
module.exports = function sub_plugin(md, abbrDefList, listPriorsFile) {
  var escapeRE        = md.utils.escapeRE,
      arrayReplaceAt  = md.utils.arrayReplaceAt;

  // ASCII characters in Cc, Sc, Sm, Sk categories we should terminate on;
  // you can check character classes here:
  // http://www.unicode.org/Public/UNIDATA/UnicodeData.txt
  var OTHER_CHARS      = ' \r\n$+<=>^`|~';

  var UNICODE_PUNCT_RE = md.utils.lib.ucmicro.P.source;
  var UNICODE_SPACE_RE = md.utils.lib.ucmicro.Z.source;

  if (abbrDefList) {
    // prepend ':' to avoid conflict with Object.prototype members
    Object.keys(abbrDefList).forEach(key => {
      if (!key.startsWith(':')) {
        Object.defineProperty(abbrDefList, ':' + key,
            Object.getOwnPropertyDescriptor(abbrDefList, key));
        delete abbrDefList[key];
      }
    });    
  }

  function abbr_def(state, startLine, endLine, silent) {
    var label, title, ch, labelStart, labelEnd,
        pos = state.bMarks[startLine] + state.tShift[startLine],
        max = state.eMarks[startLine];

    if (pos + 2 >= max) { return false; }

    if (state.src.charCodeAt(pos++) !== 0x2A/* * */) { return false; }
    if (state.src.charCodeAt(pos++) !== 0x5B/* [ */) { return false; }

    labelStart = pos;

    for (; pos < max; pos++) {
      ch = state.src.charCodeAt(pos);
      if (ch === 0x5B /* [ */) {
        return false;
      } else if (ch === 0x5D /* ] */) {
        labelEnd = pos;
        break;
      } else if (ch === 0x5C /* \ */) {
        pos++;
      }
    }

    if (labelEnd < 0 || state.src.charCodeAt(labelEnd + 1) !== 0x3A/* : */) {
      return false;
    }

    if (silent) { return true; }

    label = state.src.slice(labelStart, labelEnd).replace(/\\(.)/g, '$1');
    title = state.src.slice(labelEnd + 2, max).trim();
    if (label.length === 0) { return false; }
    if (title.length === 0) { return false; }
    if (!state.env.abbreviations) { state.env.abbreviations = {}; }
    // prepend ':' to avoid conflict with Object.prototype members
    if (typeof state.env.abbreviations[':' + label] === 'undefined') {
      state.env.abbreviations[':' + label] = title;
    }

    state.line = startLine + 1;
    return true;
  }


  function abbr_replace(state) {
    var i, j, l, tokens, token, text, nodes, pos, reg, m, regText, regSimple,
        currentToken,
        blockTokens = state.tokens;

    if (!state.env.abbreviations && !abbrDefList) {
      // no defs at all
      return;
    }
    else if (!state.env.abbreviations && abbrDefList) {
      // use specified list 
      state.env.abbreviations = abbrDefList; 
    }
    else if (state.env.abbreviations && abbrDefList) {
      // merge file defs with list defs based on priority
      state.env.abbreviations = (listPriorsFile) 
        ? Object.assign(state.env.abbreviations, abbrDefList) // list defs prior file defs
        : Object.assign(abbrDefList, state.env.abbreviations); // file defs prior list defs
    }

    regSimple = new RegExp('(?:' +
      Object.keys(state.env.abbreviations).map(function (x) {
        return x.substr(1);
      }).sort(function (a, b) {
        return b.length - a.length;
      }).map(escapeRE).join('|') +
    ')');

    regText = '(^|' + UNICODE_PUNCT_RE + '|' + UNICODE_SPACE_RE +
                    '|[' + OTHER_CHARS.split('').map(escapeRE).join('') + '])'
            + '(' + Object.keys(state.env.abbreviations).map(function (x) {
                      return x.substr(1);
                    }).sort(function (a, b) {
                      return b.length - a.length;
                    }).map(escapeRE).join('|') + ')'
            + '($|' + UNICODE_PUNCT_RE + '|' + UNICODE_SPACE_RE +
                    '|[' + OTHER_CHARS.split('').map(escapeRE).join('') + '])';

    reg = new RegExp(regText, 'g');

    for (j = 0, l = blockTokens.length; j < l; j++) {
      if (blockTokens[j].type !== 'inline') { continue; }
      tokens = blockTokens[j].children;

      // We scan from the end, to keep position when new tags added.
      for (i = tokens.length - 1; i >= 0; i--) {
        currentToken = tokens[i];
        if (currentToken.type !== 'text') { continue; }

        pos = 0;
        text = currentToken.content;
        reg.lastIndex = 0;
        nodes = [];

        // fast regexp run to determine whether there are any abbreviated words
        // in the current token
        if (!regSimple.test(text)) { continue; }

        while ((m = reg.exec(text))) {
          if (m.index > 0 || m[1].length > 0) {
            token         = new state.Token('text', '', 0);
            token.content = text.slice(pos, m.index + m[1].length);
            nodes.push(token);
          }

          token         = new state.Token('abbr_open', 'abbr', 1);
          token.attrs   = [ [ 'title', state.env.abbreviations[':' + m[2]] ] ];
          nodes.push(token);

          token         = new state.Token('text', '', 0);
          token.content = m[2];
          nodes.push(token);

          token         = new state.Token('abbr_close', 'abbr', -1);
          nodes.push(token);

          reg.lastIndex -= m[3].length;
          pos = reg.lastIndex;
        }

        if (!nodes.length) { continue; }

        if (pos < text.length) {
          token         = new state.Token('text', '', 0);
          token.content = text.slice(pos);
          nodes.push(token);
        }

        // replace current node
        blockTokens[j].children = tokens = arrayReplaceAt(tokens, i, nodes);
      }
    }
  }

  md.block.ruler.before('reference', 'abbr_def', abbr_def, { alt: [ 'paragraph', 'reference' ] });

  md.core.ruler.after('linkify', 'abbr_replace', abbr_replace);
};
